text/3730-nit-lint-level.md

```diff
@@ -0,0 +1,378 @@
+  - If stabilize with no `nit`s or only downgrading `warn`s to `nit`s, then there will be no difference in the amount of output
+  - We already have a mechanism, let's not invent a new one
+- `nits` lint group is created to give rustc-only or third-party build tools a built-in way to control these besides ignoring them like Cargo
+- Cargo could print a message that `nit`s are present to help raise awareness of how to show them but they will likely always be present, so this is noise to experienced users.
+
+# Prior art
+[prior-art]: #prior-art
+
+IntelliJ IDEA has the following [lint levels](https://www.jetbrains.com/help/idea/configuring-inspection-severities.html):
+- Error: Syntax errors
+- Warning: Code fragments that might produce bugs or require enhancement
+- Weak Warning: Code fragments that can be improved or optimized (redundant code, duplicated code fragments, and so on)
+- Server Problem
+- Grammar Error
+- Typo
+- Consideration: Code fragments that can be improved. This severity level is not marked on the error stripe and does not have a default highlighting style, but you can select one from the list of existing styles or configure your own.
+- No highlighting (but fix available)
+
+LSP has the following [diagnostic levels](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnostic):
+- Error
+- Warning
+- Information
+- Hint
+
+with the following diagnostic tags:
+- Unnecessary
+- Deprecated
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- `CARGO_BUILD_NITS=nit` is modeled off of `CARGO_BUILD_WARNINGS=warn` which is just saying "treat X level violations as Y level"
+  - `CARGO_BUILD_WARNINGS=warn` is "weird" but shouldn't show up too often as its the default
+  - However, the frequency is swapped with `CARGO_BUILD_NITS=nit`
+  - We may also want the terms to be consistent across the different level-control fields
+  - Reminder: `CARGO_BUILD_WARNINGS` is unstable and we can still change it
+
+## Naming
+
+The name should clearly communicate that there is no authoriative weight
+encouraging them to resolve them but that it is left to them and their reviewer
+to weigh whether to resolve or ignore them.
+
+The verb form is used for:
+- attribute
+- long flag
+
+Lint names are expected to be read with their level for what should be done, e.g.
+- `deny(let_and_return)` -> "deny the use of let-and-return in this code"
+- `warn(let_and_return)` -> "warn on the use of let-and-return in this code"
+
+The noun form is used for:
+- Dynamic lint group
+
+Options
+- `#[nit]` / `--nit` / `-N` / `-Wnits`
+  - Unsure how to word `nit(let_and_return)` as a sentence like the other lint levels
+- `#[note]` / `--note` / `-N` / `-Wnotes`
+  - "note the let-and-return in this code"
+- `#[notice]` / `--notice` / `-N` / `-Wnotices`
+  - "notice the let-and-return in this code"
+- `#[mention]` / `--mention` / `-M` / `-Wmentions`
+  - "mention the let-and-return in this code"
```

Comment:
@clarfonthey 

> Not to focus too much on the name, but `mention` might be another option for it. It goes well with the ordering of deny, warn, mention, allow.

Reply:
This would likely be the case for any standard communication word (`note`, `notice`, `inform`).

In part, it depends on how you frame it as rustc does communucate these by default and there are recommended workflows where these do get communicated.

`consider` seems immune *but* it has a problem with the assumed short-flag is already in use.

`hint` depends on how you word it, see https://github.com/rust-lang/rfcs/pull/3730/files#r1848937892 for more.

Reply:
Looking at [synonyms of `consider`](https://www.thesaurus.com/browse/consider)
- A non-serious option is `contemplate` :)
- `scrutinize` might work
- `inspect might work

Reply:
`info` seems rather nice along several axes: it reads reasonably well if you write `info(some_lint_name)`, it describes the concept of being informational, and it doesn't imply how it'll be displayed.

Reply:
I'm +1 on `info` (although I'm partial to `note` as it already has established meaning in the output we have today, but that might be a double edge sword if some `note`s can be controlled by the user and others can't, but we already have that for lints in general with `warning` and `error`...)

Reply:
`info` would also be parallel to Rust's logging ecosystem, which generally uses it for the [log level](https://docs.rs/log/latest/log/enum.Level.html) below `warn`. Log levels aren't a perfect analogue of lint levels, but this parallelism could still be good.

I'd also be in favor of `note`.

Reply:
existing lints/errors have help and note text. we can at least pick a new name.

Reply:
My concern with `info` is that it is a noun and  runs into problems with the level/lint naming rules at https://rustc-dev-guide.rust-lang.org/diagnostics.html#lint-naming

> The basic rule is: the lint name should make sense when read as "allow lint-name" or "allow lint-name items". For example, "allow deprecated items" and "allow dead_code" makes sense, while "allow unsafe_block" is ungrammatical (should be plural).


This is why in the RFC body I switched the level name to `inform`.

Reply:
I'll point out that nitpick is also a verb:

> find or point out minor faults in a fussy or pedantic way.

Pedantic would also seem appropriate, but that is an adjective, not a verb, and the verb form "pedantize" is unusual, and doesn't quite fit.